libsolidity/interface/CompilerStack.cpp

```diff
@@ -103,6 +107,98 @@ CompilerStack::~CompilerStack()
 	TypeProvider::reset();
 }
 
+void CompilerStack::createAndAssignCallGraphs()
+{
+	for (Source const* source: m_sourceOrder)
+		if (source->ast)
+			for (ASTPointer<ASTNode> const& node: source->ast->nodes())
+				if (auto const* contract = dynamic_cast<ContractDefinition*>(node.get()))
+				{
+					Contract& contractState = m_contracts.at(contract->fullyQualifiedName());
+
+					contractState.contract->annotation().creationCallGraph = make_unique<CallGraph>(
+						FunctionCallGraphBuilder::buildCreationGraph(
+							*contract
+						)
+					);
+					contractState.contract->annotation().deployedCallGraph = make_unique<CallGraph>(
+						FunctionCallGraphBuilder::buildDeployedGraph(
+							*contract,
+							**contractState.contract->annotation().creationCallGraph
+						)
+					);
+
+					auto contractDeps = ranges::views::concat(
+						contract->annotation().creationCallGraph->get()->bytecodeDependency,
+						contract->annotation().deployedCallGraph->get()->bytecodeDependency
+					);
+
+					for (auto const& [dependencyContract, referencee]: contractDeps)
+					{
+						solAssert(dependencyContract != nullptr, "");
+						contractState.contract->annotation().contractDependencies.emplace(dependencyContract, referencee);
+					}
+				}
+}
+
+void CompilerStack::findAndReportCyclicContractDependencies()
```

Comment:
Why does this look so much more complicated than ConstStateVarCircularReferenceChecker ?

Reply:
One of the factors that makes this more complicated is this: For the const variables, we only have the declaration and the identifier that causes a cycling dependency.
For contracts, we start with the contract and find a dependency cycle. But then we also want to know which ASTNode (expression) caused the dependency and that information is also in the map `contractDependencies` but it might not be in the originating contract but instead in a dependent contract. 
So to find the ast node that made the reference, I either have to go through all dependencies (and their dependencies) again, finding the contract -> ast node mapping or I remember it and save it when I find the cycle. The later thing is what I am trying to do, but it is very difficult to do this cleanly. I tried to avoid modifying or extending the CycleDetector directly, but that leaves me only limited options, e.g. using the `RememberingCycleDetector` and when I investigated some parts here yesterday, I started to suspect that it's not even done right at the moment so I might have to find another solution to do this. Maybe I have to extend it directly...